README.md
# storj-mongodb

To build from scratch, [install Go](https://golang.org/doc/install#install).

```
go get github.com/utropicmedia/storj-mongodb
```

## Set-up Files
* Create a `db_property.json` file, with following contents about a MongoDB instance:
    * hostName :- Host Name connect to MongoDB
    * port :- Port connect to MongoDB
    * username :- User Name of MongoDB
    * password :- Password of MongoDB
    * database :- MongoDB Database Name

```json
    { 
        "hostname": "mongodbHostName",
        "port":     "27017",
        "username": "username",
        "password": "password",
        "database": "mongoDatabaseName"
    }
```

* Create a `storj_config.json` file, with Storj network's configuration information in JSON format:
    * apiKey :- API key created in Storj satellite gui
    * satelliteURL :- Storj Satellite URL
    * encryptionPassphrase :- Storj Encryption Passphrase.
    * bucketName :- Split file into given size before uploading.
    * uploadPath :- Path on Storj Bucket to store data (optional) or "/"
    * serializedScope:- Serialized Scope Key shared while uploading data used to access bucket without API key
    * disallowReads:- Set true to create serialized scope key with restricted read access
    * disallowWrites:- Set true to create serialized scope key with restricted write access
    * disallowDeletes:- Set true to create serialized scope key with restricted delete access

```json
    { 
        "apikey":     "change-me-to-the-api-key-created-in-satellite-gui",
        "satellite":  "us-central-1.tardigrade.io:7777",
        "bucket":     "change-me-to-desired-bucket-name",
        "uploadPath": "optionalpath/requiredfilename",
        "encryptionpassphrase": "you'll never guess this",
        "serializedScope": "change-me-to-the-api-key-created-in-encryption-access-apiKey",
        "disallowReads": "true/false-to-disallow-reads",
        "disallowWrites": "true/false-to-disallow-writes",
        "disallowDeletes": "true/false-to-disallow-deletes"
    }
```

* Store both these files in a `config` folder.  Filename command-line arguments are optional.  defualt locations are used.

## Run the command-line tool

* Import the package "github.com/utropicmedia/storj-mongodb" in your main file and call the funtion 'strojmongodb.StorjMongoDB()' in the main function.

**NOTE**: The following commands operate in a Linux system.

* Get help
```
$ storj-mongodb -h
```

* Check version
```
$ storj-mongodb -v
```

* Read BSON data from desired MongoDB instance and upload it to given Storj network bucket using Serialized Scope Key.  [note: filename arguments are optional.  default locations are used.]
```
$ storj-mongodb store ./config/db_property.json ./config/storj_config.json  
```

* Read BSON data from desired MongoDB instance and upload it to given Storj network bucket API key and EncryptionPassPhrase from storj_config.json and creates an unrestricted shareable Serialized Scope Key.  [note: filename arguments are optional. default locations are used.]
```
$ storj-mongodb store ./config/db_property.json ./config/storj_config.json key
```

* Read BSON data from desired MongoDB instance and upload it to given Storj network bucket API key and EncryptionPassPhrase from storj_config.json and creates a restricted shareable Serialized Scope Key.  [note: filename arguments are optional. default locations are used. `restrict` can only be used with `key`]
```
$ storj-mongodb store ./config/db_property.json ./config/storj_config.json key restrict
```

* Read BSON data in `debug` mode from desired MongoDB instance and upload it to given Storj network bucket.  [note: filename arguments are optional.  default locations are used. Make sure `debug` folder already exist in project folder.]
```
$ storj-mongodb store debug ./config/db_property.json ./config/storj_config.json  
```

* Read MongoDB instance property from a desired JSON file and display all its collections' data
```
$ storj-mongodb parse   
```

* Read MongoDB instance property in `debug` mode from a desired JSON file and display all its collections' data
```
$ storj-mongodb parse debug 
```

* Read and parse Storj network's configuration, in JSON format, from a desired file and upload a sample object
```
$ storj-mongodb test 
```
* Read and parse Storj network's configuration, in JSON format, from a desired file and upload a sample object in `debug` mode
```
$ storj-mongodb test debug 
```


CHANGELOG.md
# Storj-MongoDB Changelog

## [1.0.15] - 12-04-2020
### Changelog:
* Resolved the syntax errors occured during build command.

## [1.0.14] - 13-02-2020
### Changelog:
* Changed paths of the go files and corresponding packages.


## [1.0.13] - 12-02-2020
### Changelog:
* Fixed the module related issues by changing the path of the storjmongodb module.
* Paths of mongo and storj packages have also been altered by removing the mongodb_storj_inteface folder.

## [1.0.12] - 07-02-2020
### Changelog:
* UserAgent name changed to MongoDB.
* Made storj-mongodb a package named storjmongodb.

## [1.0.11] - 05-02-2020
### Changelog:
* Changes made accodring to latest uplink v0.31.13.
* Updated uplink, storj and other libraries.
* Simplified code structure.
* Removed unneeded aliases in mongo.go.
* Fixed path handeling in stroj.go.
* Added a downloadObject fucntion in storj.go
* Replaced partnerID with userAgent and named as MongoDB.

## [1.0.10] - 17-12-2019
### Changelog:
* Changes made according to latest libuplink v0.27.1
* Changes made according to updated cli package.
* Added Macroon functionality.
* Added option to access storj using Serialized Scope Key. 
* Added keyword `key` to access storj using API key rather than Serialized Scope Key (defalt).
* Added keyword `restrict` to apply restrictions on API key and provide shareable Serialized Scope Key for users.
* Error handling for various events.


mongodb_storj_interface/storj/storj.go
// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

package storj

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"storj.io/storj/lib/uplink"
	"storj.io/storj/pkg/macaroon"
)

// DEBUG allows more detailed working to be exposed through the terminal.
var DEBUG = false

// ConfigStorj depicts keys to search for within the stroj_config.json file.
type ConfigStorj struct {
	APIKey               string `json:"apikey"`
	Satellite            string `json:"satellite"`
	Bucket               string `json:"bucket"`
	UploadPath           string `json:"uploadPath"`
	EncryptionPassphrase string `json:"encryptionpassphrase"`
	SerializedScope      string `json:"serializedScope"`
	DisallowReads        string `json:"disallowReads"`
	DisallowWrites       string `json:"disallowWrites"`
	DisallowDeletes      string `json:"disallowDeletes"`
}

// LoadStorjConfiguration reads and parses the JSON file that contain Storj configuration information.
func LoadStorjConfiguration(fullFileName string) (ConfigStorj, error) { // fullFileName for fetching storj V3 credentials from  given JSON filename.

	var configStorj ConfigStorj

	fileHandle, err := os.Open(fullFileName)
	if err != nil {
		return configStorj, err
	}
	defer fileHandle.Close()

	jsonParser := json.NewDecoder(fileHandle)
	jsonParser.Decode(&configStorj)

	// Display read information.
	fmt.Println("\nRead Storj configuration from the ", fullFileName, " file")
	fmt.Println("\nAPI Key\t\t: ", configStorj.APIKey)
	fmt.Println("Satellite	: ", configStorj.Satellite)
	fmt.Println("Bucket		: ", configStorj.Bucket)
	fmt.Println("Upload Path\t: ", configStorj.UploadPath)
	fmt.Println("Serialized Scope Key\t: ", configStorj.SerializedScope)

	return configStorj, nil
}

// ConnectStorjReadUploadData reads Storj configuration from given file,
// connects to the desired Storj network.
// It then reads data using io.Reader interface and
// uploads it as object to the desired bucket.
func ConnectStorjReadUploadData(fullFileName string, databaseReader io.Reader, databaseName string, keyValue string, restrict string) (string, error) { // fullFileName for fetching storj V3 credentials from  given JSON filename
	// databaseReader is an io.Reader implementation that 'reads' desired data,
	// which is to be uploaded to storj V3 network.
	// databaseName for adding dataBase name in storj V3 filename.
	// Read Storj bucket's configuration from an external file.
	var scope string
	configStorj, err := LoadStorjConfiguration(fullFileName)
	if err != nil {
		log.Fatal("loadStorjConfiguration:", err)
	}

	fmt.Println("\nCreating New Uplink...")

	var cfg uplink.Config
	// Configure the partner id
	cfg.Volatile.UserAgent = "MongoDB"

	ctx := context.Background()

	uplinkstorj, err := uplink.NewUplink(ctx, &cfg)
	if err != nil {
		uplinkstorj.Close()
		log.Fatal("Could not create new Uplink object:", err)
	}
	defer uplinkstorj.Close()
	var serializedScope string
	if keyValue == "key" {

		fmt.Println("Parsing the API key...")
		key, err := uplink.ParseAPIKey(configStorj.APIKey)
		if err != nil {
			uplinkstorj.Close()
			log.Fatal("Could not parse API key:", err)
		}

		if DEBUG {
			fmt.Println("API key \t   :", configStorj.APIKey)
			fmt.Println("Serialized API key :", key.Serialize())
		}

		fmt.Println("Opening Project...")
		proj, err := uplinkstorj.OpenProject(ctx, configStorj.Satellite, key)
		if err != nil {
			uplinkstorj.Close()
			log.Fatal("Could not open project:", err)
		}
		defer proj.Close()

		// Creating an encryption key from encryption passphrase.
		if DEBUG {
			fmt.Println("\nGetting encryption key from pass phrase...")
		}

		encryptionKey, err := proj.SaltedKeyFromPassphrase(ctx, configStorj.EncryptionPassphrase)
		if err != nil {
			uplinkstorj.Close()
			proj.Close()
			log.Fatal("Could not create encryption key:", err)
		}

		// Creating an encryption context.
		access := uplink.NewEncryptionAccessWithDefaultKey(*encryptionKey)
		if DEBUG {
			fmt.Println("Encryption access \t:", configStorj.EncryptionPassphrase)
		}

		// Serializing the parsed access, so as to compare with the original key.
		serializedAccess, err := access.Serialize()
		if err != nil {
			uplinkstorj.Close()
			proj.Close()
			log.Fatal("Error Serialized key : ", err)
		}

		if DEBUG {
			fmt.Println("Serialized access key\t:", serializedAccess)
		}

		// Load the existing encryption access context
		accessParse, err := uplink.ParseEncryptionAccess(serializedAccess)
		if err != nil {
			log.Fatal(err)
		}

		if restrict == "restrict" {
			disallowRead, _ := strconv.ParseBool(configStorj.DisallowReads)
			disallowWrite, _ := strconv.ParseBool(configStorj.DisallowWrites)
			disallowDelete, _ := strconv.ParseBool(configStorj.DisallowDeletes)
			userAPIKey, err := key.Restrict(macaroon.Caveat{
				DisallowReads:   disallowRead,
				DisallowWrites:  disallowWrite,
				DisallowDeletes: disallowDelete,
			})
			if err != nil {
				log.Fatal(err)
			}
			userAPIKey, userAccess, err := accessParse.Restrict(userAPIKey,
				uplink.EncryptionRestriction{
					Bucket:     configStorj.Bucket,
					PathPrefix: configStorj.UploadPath,
				},
			)
			if err != nil {
				log.Fatal(err)
			}
			userRestrictScope := &uplink.Scope{
				SatelliteAddr:    configStorj.Satellite,
				APIKey:           userAPIKey,
				EncryptionAccess: userAccess,
			}
			serializedRestrictScope, err := userRestrictScope.Serialize()
			if err != nil {
				log.Fatal(err)
			}
			scope = serializedRestrictScope
		}
		userScope := &uplink.Scope{
			SatelliteAddr:    configStorj.Satellite,
			APIKey:           key,
			EncryptionAccess: access,
		}
		serializedScope, err = userScope.Serialize()
		if err != nil {
			log.Fatal(err)
		}
		if restrict == "" {
			scope = serializedScope
		}

		proj.Close()
		uplinkstorj.Close()
	} else {
		serializedScope = configStorj.SerializedScope

	}
	parsedScope, err := uplink.ParseScope(serializedScope)
	if err != nil {
		log.Fatal(err)
	}

	uplinkstorj, err = uplink.NewUplink(ctx, &cfg)
	if err != nil {
		log.Fatal("Could not create new Uplink object:", err)
	}
	proj, err := uplinkstorj.OpenProject(ctx, parsedScope.SatelliteAddr, parsedScope.APIKey)
	if err != nil {
		uplinkstorj.Close()
		proj.Close()
		log.Fatal("Could not open project:", err)
	}

	fmt.Println("Opening Bucket: ", configStorj.Bucket)

	// Open up the desired Bucket within the Project.
	bucket, err := proj.OpenBucket(ctx, configStorj.Bucket, parsedScope.EncryptionAccess)
	if err != nil {
		fmt.Println("Could not open bucket", configStorj.Bucket, ":", err)
		fmt.Println("Trying to create new bucket....")
		_, err1 := proj.CreateBucket(ctx, configStorj.Bucket, nil)
		if err1 != nil {
			uplinkstorj.Close()
			proj.Close()
			bucket.Close()
			fmt.Printf("Could not create bucket %q:", configStorj.Bucket)
			log.Fatal(err1)
		} else {
			fmt.Println("Created Bucket", configStorj.Bucket)
		}
		fmt.Println("Opening created Bucket: ", configStorj.Bucket)
		bucket, err = proj.OpenBucket(ctx, configStorj.Bucket, parsedScope.EncryptionAccess)
		if err != nil {
			fmt.Printf("Could not open bucket %q: %s", configStorj.Bucket, err)
		}
	}

	defer bucket.Close()

	var fileNamesDEBUG []string
	checkSlash := configStorj.UploadPath[len(configStorj.UploadPath)-1:]
	if checkSlash != "/" {
		configStorj.UploadPath = configStorj.UploadPath + "/"
	}

	// Read data using io.Reader and upload it to Storj.
	for err = io.ErrShortBuffer; err == io.ErrShortBuffer; {
		t := time.Now()
		timeNow := t.Format("2006-01-02_15:04:05")
		var filename = databaseName + "/" + timeNow + ".bson"
		//
		fmt.Println("File path: ", configStorj.UploadPath+filename)
		fmt.Println("\nUploading of the object to the Storj bucket: Initiated...")

		err = bucket.UploadObject(ctx, configStorj.UploadPath+filename, databaseReader, nil)
		if DEBUG {
			fileNamesDEBUG = append(fileNamesDEBUG, filename)

		}
	}

	if err != nil {
		fmt.Printf("Could not upload: %s\t", err)
		return scope, err
	}

	fmt.Println("Uploading of the object to the Storj bucket: Completed!")

	if DEBUG {
		for _, filename := range fileNamesDEBUG {
			// Test uploaded data by downloading it.
			// serializedAccess, err := access.Serialize().
			// Initiate a download of the same object again.
		
			fmt.Printf("Downloading Object %s from bucket : Initiated...\n", filename)
			// Read everything from the stream.
			receivedContents, err := downloadObject(ctx, bucket, configStorj.UploadPath + filename)
			if err != nil {
				return scope, fmt.Errorf("could not download object: %v", err)
			}
			var decodedBson bson.M
			bson.Unmarshal(receivedContents, &decodedBson)
			_, err = json.Marshal(decodedBson)
			if err != nil {
				panic(err) // it will be invoked
				// panic: json: unsupported value: NaN
			}
			path := strings.Split(filename, "/")

			pathtokens := strings.Split(path[1], ":")
		_ = os.MkdirAll(filepath.Join("debug", path[0]), 0644)
			var fileNameDownload = filepath.Join("debug", path[0], pathtokens[0], pathtokens[1], pathtokens[2])

			err = ioutil.WriteFile(fileNameDownload, receivedContents, 0644)
			if err != nil {
				fmt.Println(err)
			}

			fmt.Printf("Downloaded %d bytes of Object from bucket!\n", len(receivedContents))
		}
	}

	return scope, nil
}

func downloadObject(ctx context.Context, bucket *uplink.Bucket, path string) ([]byte, error) {
	strm, err := bucket.Download(ctx, path)
	if err != nil {
		return nil, fmt.Errorf("could not open object at %q: %v", path, err)
	}
	defer strm.Close()

	// Read everything from the stream.
	receivedContents, err := ioutil.ReadAll(strm)
	if err != nil {
		return nil, fmt.Errorf("could not read object: %v", err)
	}

	return receivedContents, err
}